buildroot/share/scripts/upload.py

```diff
@@ -0,0 +1,193 @@
+        print(f" -Target      : {upload_firmware_target_name}")
+        print(f" -Port        : {upload_port} @ {upload_speed} baudrate")
+        print(f" -Timeout     : {upload_timeout}")
+        print(f" -Block size  : {upload_blocksize}")
+        print(f" -Compression : {upload_compression}")
+        print(f" -Error ratio : {upload_error_ratio}")
+        print(f" -Test        : {upload_test}")
+        print(f" -Reset       : {upload_reset}")
+        print("----------------")
+
+    # Init serial port
+    port = serial.Serial(upload_port, baudrate = upload_speed, write_timeout = 0, timeout = 0.1)
+    port.reset_input_buffer()
+
+    # Check SD card status
+    if not _CheckSDCard(): exit(2)
+    
+    # Get firmware files
+    (Status, FirmwareFiles) = _GetFirmwareFiles()
+    if not Status: exit(3)
+    if Debug:
+        for FirmwareFile in FirmwareFiles:
+            print("Found: " + FirmwareFile)
+        
+    # Get all 1st level firmware files (to remove)
+    OldFirmwareFiles = _FilterFirmwareFiles(FirmwareFiles[1:len(FirmwareFiles)-2])   # Skip header and footers of list
+    if len(OldFirmwareFiles) == 0:
+        print("No old firmware files found")
+    else:
+        print(f"Remove {len(OldFirmwareFiles)} old firmware file{'s' if len(OldFirmwareFiles) != 1 else ''}:")
+        for OldFirmwareFile in OldFirmwareFiles:
+            print(" -Removing- " + OldFirmwareFile + "...", end='', flush=True)
+            print("OK" if _RemoveFirmwareFile(OldFirmwareFile) else "Error!")
+
+    # Cleanup completed
+    port.close()
+    if Debug: print("Cleanup completed")
+    
+    try:
+        # Upload File
+        print("Copying '" + upload_firmware_source_name + "' --> '" + upload_firmware_target_name + "'")
+        protocol = MarlinBinaryProtocol.Protocol(upload_port, upload_speed, upload_blocksize, float(upload_error_ratio), int(upload_timeout))
+        #echologger = MarlinBinaryProtocol.EchoProtocol(protocol)
+        protocol.connect()
+        filetransfer = MarlinBinaryProtocol.FileTransferProtocol(protocol)
+        filetransfer.copy(upload_firmware_source_name, upload_firmware_target_name, upload_compression, upload_test)
+        protocol.disconnect()
+
+        # Notify upload completed
+        protocol.send_ascii("M117 Firmware uploaded")
+
+        # Remount SD card
+        print('Wait for SD card release...')
+        time.sleep(1)
+        print('Remount SD card')
+        protocol.send_ascii("M21")
+
+        if upload_reset:
+            print("Trigger firmware update...")
+            protocol.send_ascii("M997", True)
+    except KeyboardInterrupt:
+        if filetransfer: filetransfer.abort()
+
+    except MarlinBinaryProtocol.FatalError:
+        print("Too Many Retries, Abort")
+
+    finally:
+        if protocol: protocol.shutdown()
+        print("Firmware upload completed")
+    
+
+# Attach custom upload callback
+env.Replace(UPLOADCMD=Upload)
```

Comment:
This was the line that inspired my earlier comment. At the time I hadn't read up on how to implement a custom upload command in PlatformIO.

Is this going to replace the upload command, no matter what is specified for `upload_protocol`, or will it only replace if `upload_protocol=custom` is specified? It would be nice if we can always register the custom handler, but still work properly if the chooses to use `upload_protocol=stlink`, or other protocols available from the platform.

I was hoping that it would be be possible for us to implement multiple custom upload_protocols. I was thinking of something like this as a possible future, where people could uncomment the one they want to use. Some would of course need additional options to be useful.

```
#upload_protocol = mf_binary_protocol   # Your implementation
#upload_protocol = mf_usb_mass_storage  # The LPC1768 platform already implements this by default
#upload_protocol = mf_octoprint         # I'd love to forward the binary to my OctoPi to do the update for me!
```

Being able to treat each as a top-level protocol felt cleaner to me than having a single "custom" upload with lots of parameters, but reading PlatformIO docs it doesn't seem obvious that could be done outside of the platform definition.

Reply:
From a quick look at the pio documentation I can't find info on custom "upload_protocol", but we can use custom project parameters to achieve something similar:
```
[env: ... ]
upload_protocol  = custom
upload_target    = mf_binary_protocol   # Your implementation
#upload_target   = mf_usb_mass_storage  # The LPC1768 platform already implements this by default
#upload_target   = mf_octoprint         # I'd love to forward the binary to my OctoPi to do the update for me!
#upload_filename = firmware.bin         # If a filename override is needed
```